pilot/pkg/bootstrap/istio_ca.go

```diff
@@ -227,23 +230,22 @@ func detectSigningCABundle() (ca.SigningCAFileBundle, error) {
 	tlsSigningFile := path.Join(LocalCertDir.Get(), ca.TLSSecretCACertFile)
 
 	// looking for tls file format (tls.crt)
+	// If tls.crt is used - the cert may include intermediaries, but not the root cert(s)
+	//
 	if _, err := os.Stat(tlsSigningFile); err == nil {
-		log.Info("Using kubernetes.io/tls secret type for signing ca files")
+		log.Infof("Using kubernetes.io/tls secret type for signing ca files %s", LocalCertDir.Get())
 		return ca.SigningCAFileBundle{
-			RootCertFile: path.Join(LocalCertDir.Get(), ca.TLSSecretRootCertFile),
-			CertChainFiles: []string{
-				tlsSigningFile,
-				path.Join(LocalCertDir.Get(), ca.TLSSecretRootCertFile),
-			},
+			RootCertFile:   path.Join(LocalCertDir.Get(), ca.TLSSecretRootCertFile),
+			CertChainFiles: nil,
```

Comment:
This is the case ( I hope ) where tls.crt is used - which has the leaf certificate plus all intermediaries ( the CertChain). So it is already concatenated in SigningCertFile - however to verify it is valid we need to split it there ( in the validation code ), because the validator expects one single-cert PEM as leaf, and separate intermediaryes.

We certainly don't want to concatenate the tlsSigningFile - that includes the leaf plus all certs - and the roots.

Reply:
given we already used to read tls.crt and appended, are we sure this isn't going to break users? 

IIUC this is changing the contract; before we concate and now we don't. 

I think I know *why*,but if there were >users of tls.crt won't they be broken?

Reply:
Besides comments and better method names - this is really what I've been trying to fix. I think users are broken now ( the bug about ztunnel with interediary certs), because tls.crt was not handled the same way as the old style certs. 

The rest of the code expects one leaf in a file, the set of intermediaries in another - and the roots at the end. When tls.crt was added, it no longer had this form - if I remember correctly from 1 month ago debugging, there were more certs in the chain because of concatenation - somehow that passed the checks in Istiod ( because the validation stops when a root is reached), but rust may be more strict. Or at least that was my theory - I've been using ztunnel with the intermediary certs and this patch so it may be working.

Reply:
Am I wrong to think there is no specified format for how "tls.crt" should behave? So we weren't doing it "wrong" before we just had a different expectation of the unspecified format? 

If so it seems like we are fixing those users while breaking other users? 

If not, where is the specification

Reply:
The bug reports on ztunnel suggest we were doing it 'wrong' - and it may have worked with Envoy because it's more tolerant with the cert chains.

How tls.crt should behave is pretty clear - it is the full chain, starting with the leaf (real cert) and followed by the list of intermediaries, not including the roots. K8S, CertManager - and most of the internet - is using this model.

The Istio model was to have the leaf in one file, the intermediaries in another - and there are some assumptions about this.

Validating the chain is a tricky business - because there is a feature to allow 'cross-signing' ( ACME is a good example), the process can be:
- check every cert in the chain for validity - keeping track of "Subject" and "Issuer"
- try to form chains following Issuer-Subject until you reach one of the trusted CAs ( which are indexed by Subject)

or it can be more strict:
- expect an ordered list of certs, each cert signed by the next, with Issuer of the previous cert matching Subject of the next cert.

There are other variations - java had its own way, browsers had different behaviors - but the most reliable and safe way is the second one, and that's what the original Istio did. Some validators tolerate out of order certs in the chain, some don't.

I suspect with the implementation of tls.crt - we end up with multiple intermediates. Or at least that's what I remember from 1 month ago when I wrote the PR :-)

Reply:
The getter in security/pkg/pki/util/keycertbundle.go is widely used, i think it will make KeyCertBundle.GetCertChainPem return empty

Reply:
> How tls.crt should behave is pretty clear - it is the full chain, starting with the leaf (real cert) and followed by the list of intermediaries, not including the roots. K8S, CertManager - and most of the internet - is using this model.
> The Istio model was to have the leaf in one file, the intermediaries in another - and there are some assumptions about this.

Is it feasible to concurrently support both and detect? if not, it seems like we would need some flag (and clear error logs if they are not set).

IIUC `tls.crt` is just a file. There are two competing "standards" to configure this, so we cannot just kill one of them, no?